quil-rs/src/expression/simplification/by_hand.rs

```diff
@@ -17,74 +24,109 @@ pub(super) fn run(expression: &Expression) -> Expression {
 /// crash with an "I've overflowed my stack" error. Except for exceedingly large expressions
 /// (`the_big_one` test case in `mod.rs`, for example), a larger limit here doesn't seem to be of
 /// practical value in anecdotal testing.
-const LIMIT: u64 = 10;
-
-/// Recursively simplify an [`Expression`] by hand, breaking into cases to make things more
-/// manageable.
-fn simplify(e: &Expression, limit: u64) -> Expression {
-    if limit == 0 {
-        // bail
-        e.clone()
-    } else {
-        match e {
+const LIMIT: Limit = Limit(10);
+
+/// We use a separate type for the limit so that we can enforce the use of saturating subtraction,
+/// but we don't use [`std::num::Saturating`] so that we can write that in terms of literals.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[repr(transparent)]
+struct Limit(u64);
+
+impl std::cmp::PartialEq<u64> for Limit {
+    #[inline(always)]
+    fn eq(&self, other: &u64) -> bool {
+        self.0 == *other
+    }
+}
+
+impl std::ops::Sub<u64> for Limit {
+    type Output = Self;
+
+    #[inline(always)]
+    fn sub(self, rhs: u64) -> Self::Output {
+        Limit(self.0.saturating_sub(rhs))
+    }
+}
+
+#[derive(Debug, Clone)]
+struct Simplifier {
+    memo_table: HashMap<ArcIntern<Expression>, ArcIntern<Expression>>,
+}
+
+impl Simplifier {
+    fn new() -> Self {
+        Self {
+            memo_table: HashMap::new(),
+        }
+    }
+
+    #[allow(dead_code)]
+    fn debug_memo_table(&self, w: &mut impl std::io::Write) -> std::io::Result<()> {
```

Comment:
does/did this get used for tests?

Reply:
Yes, while I was making sure I'd written the code right – I'll add a comment to that effect